src/main/java/org/quiltmc/loader/api/config/Config.java

```diff
@@ -0,0 +1,420 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.api.config;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Consumer;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.quiltmc.loader.api.config.values.ValueList;
+import org.quiltmc.loader.api.config.values.ValueMap;
+import org.quiltmc.loader.api.config.values.ValueTreeNode;
+import org.quiltmc.loader.impl.config.ConfigImpl;
+
+@ApiStatus.NonExtendable
+public interface Config {
+	String getModId();
+
+	String getId();
+
+	/**
+	 * The path this config will be saved in, relative to Quilt Loader's config directory.
+	 *
+	 * @return a save path
+	 */
+	Path getSavePath();
+
+	/**
+	 * Adds a listener to this {@link Config} that's called whenever any of its values is updated
+	 *
+	 * @param callback an update listener
+	 */
+	void register(UpdateCallback callback);
+
+	Iterable<String> flags();
```

Comment:
Flags are another kind of metadata. Essentially, flags are unique string values that are stored in a set, while metadata has a list of metadata values for each type. I did consider making flags just a metadata type instead of its own construct, but I thought that the uniqueness might warrant keeping them around.

Reply:
I think for simplicity, Flags should be metadata. However, having these existing functions as helpers could still be nice.

Reply:
I don't think the added simplicity of having one metadata type outweighs the added complexity of consumers having to check for potential duplicates.

Reply:
Why would consumers have to check for duplicates?
Just make a Metadata for a Set of Strings for flags. Pretty sure that's the exact same functionality.